torch/csrc/jit/codegen/cuda/parser.cpp

```diff
@@ -451,6 +451,350 @@ class IrParser {
+            auto mean_hat = mul(rmean_bcast, new Float(1.0 - kMomentum));
+            auto new_mean_hat = add(mean_hat, current_mean_hat);
+            */
+
+            auto x_mean_sub = sub(input, x_mean);
+            auto x_mean_sub_pow =
+                binaryOp(BinaryOpType::Pow, x_mean_sub, new Float(2));
+            auto var_sum = sum(x_mean_sub_pow, reduction_axes);
+            auto var_sum_bcast = broadcast(var_sum, broadcast_mask);
+            auto var = div(var_sum_bcast, num_features);
+
+            /*
+            auto current_var_hat = mul(var, new Float(kMomentum));
+            auto rvar_bcast = broadcast(running_var, broadcast_mask);
+            auto var_hat = mul(rvar_bcast, new Float(1.0 - kMomentum));
+            auto new_var_hat = add(var_hat, current_var_hat);
+            */
+
+            auto var_eps = add(var, new Float(kEps));
+            auto rvar = unaryOp(UnaryOpType::Rsqrt, var_eps);
+            auto output = mul(x_mean_sub, rvar);
+
+            std::vector<TensorView*> all_tensors({input,
+                                                  // current_mean,
+                                                  // rmean_bcast,
+                                                  // mean_hat,
+                                                  // current_mean_hat,
+                                                  // current_var,
+                                                  // rvar_bcast,
+                                                  // var_hat,
+                                                  // current_var_hat,
+                                                  x_sum,
+                                                  var_sum,
+                                                  x_mean,
+                                                  x_sum_bcast,
+                                                  x_mean_sub,
+                                                  x_mean_sub_pow,
+                                                  var_sum_bcast,
+                                                  var,
+                                                  var_eps,
+                                                  rvar,
+                                                  output});
+
+            // Optional: norm * weight
+            if (weight) {
+              auto weight_bcast = broadcast(weight, broadcast_mask);
+              output = mul(output, weight_bcast);
+              all_tensors.push_back(weight_bcast);
+              all_tensors.push_back(output);
+            }
+
+            // Optional: norm * weight + bias
+            if (bias) {
+              auto bias_bcast = broadcast(bias, broadcast_mask);
+              output = add(output, bias);
+              all_tensors.push_back(bias_bcast);
+              all_tensors.push_back(output);
+            }
+            value_map.emplace(node->output()->unique(), output);
+
+            // Merge reduction axes together
```

Comment:
Looks like that we are mixing computation with scheduling. I would try to avoid this if possible.

Reply:
For scheduling, we need to merge all non-reduction and reduction axes together and to place the merged reduction axis in the -1 position.

1) Softmax - Single Axis Reduction
[N,  C,  H,  W] => (mark reduction axis) [N, rC,  H,  W] => (merge then reorder) [N, H*W, rC]

2) LayerNorm - Contiguous, Multiple Reduction Axes
[N,  C,  H,  W] => (mark reduction axes) [N, C, rH, rW] => (merge) [N*C, rH*rW]

3) BatchNorm - Non-Contiguous, Multiple Reduction Axes
[N,  C,  H,  W] => (mark reduction axes) [rN, C, rH, rW] => (reorder then merge) [C, rN*rH*rW]

If you wanted to handle this schedule.cpp, you would need to infer which case you have before transforming the TensorViews. This information is already given in parser.cpp.

Reply:
Totally agree that this is a convenient place to do that. But putting transformation in parser means that scheduler needs to be made aware of pre-schedule transformations.
I can live with us taking the short cut, but maybe we should re-name multiple reduction scheduler more specifically to normalization scheduler and restricting its application to only those operations? Things that would need to be done is to define normalization operations (expand this: https://github.com/csarofeen/pytorch/blob/52581a0bf13f1a92e8346e6fbb08c7f44fcb9d55/torch/csrc/jit/codegen/cuda/parser.cpp#L163-L166 to identify normalization, as well as updating `graph_fuser.cpp`). Alternatively we can put this into our fusion partition (if we are going that route instead). This should be done in a separate PR tho.

Side note, complications from this:
Permutation on stride in integration doesn't work with this, because we don't have reduction axes for normalization layers. This is a problem we need to solve. An open question is, should we wait until we have permutation in codegen? [I'll open an issue tagging NHWC]